File: src/typescript/src/pipeline_services/examples/example07_prompt_codegen_pipeline.ts
import { existsSync } from "node:fs";
import path from "node:path";

import {
  PipelineJsonLoader,
  PipelineRegistry,
} from "../../index.js";
import { register_generated_actions } from "../generated/index.js";
import { strip } from "./text_steps.js";

function findPipelineFile(pipelineFileName: string): string {
  let currentDir = process.cwd();
  while (true) {
    const candidatePath = path.join(
      currentDir,
      "pipelines",
      pipelineFileName,
    );
    if (existsSync(candidatePath)) {
      return candidatePath;
    }
    const parentDir = path.dirname(currentDir);
    if (parentDir === currentDir) {
      break;
    }
    currentDir = parentDir;
  }
  throw new Error(
    "Could not locate pipelines directory from current working directory",
  );
}

async function main(): Promise<void> {
  const pipelineFile = findPipelineFile("normalize_name.json");

  const registry = new PipelineRegistry();
  registry.register_unary("strip", strip);
  register_generated_actions(registry);

  const pipeline = await new PipelineJsonLoader().load_file(
    pipelineFile,
    registry,
  );
  const output = await pipeline.run("  john   SMITH ");
  // eslint-disable-next-line no-console
  console.log("output=" + String(output));
}

void main();

File: src/Python/pipeline_services/examples/example07_prompt_codegen_pipeline.py
from __future__ import annotations

from pathlib import Path

from pipeline_services.config.json_loader import PipelineJsonLoader
from pipeline_services.core.registry import PipelineRegistry
from pipeline_services.examples.text_steps import strip
from pipeline_services.generated import register_generated_actions


def find_pipeline_file(pipeline_file_name: str) -> Path:
    current_path = Path.cwd().resolve()
    while True:
        candidate_file = current_path / "pipelines" / pipeline_file_name
        if candidate_file.exists():
            return candidate_file
        parent_path = current_path.parent
        if parent_path == current_path:
            break
        current_path = parent_path
    raise FileNotFoundError("Could not locate pipelines directory from current working directory")


def main() -> None:
    pipeline_file = find_pipeline_file("normalize_name.json")

    registry = PipelineRegistry()
    registry.register_unary("strip", strip)
    register_generated_actions(registry)

    loader = PipelineJsonLoader()
    pipeline = loader.load_file(str(pipeline_file), registry)
    result = pipeline.run("  john   SMITH ")
    print("output=" + str(result.context))


if __name__ == "__main__":
    main()
